tools/generator.js

```diff
@@ -0,0 +1,163 @@
+            'contentType': mime_type,
+            'license': license,
+            'tags': [
+                'EknMediaObject'
+            ],
+            'height': height,
+            'width': width,
+            'source': path
+        };
+    },
+
+    _create_set: function (path, title, child_tags, tags, thumbnail_path) {
+        let _tags = ['EknSetObject'];
+        let featured = false;
+
+        if (tags.length === 0) {
+            _tags.push('EknHomePageTag');
+            featured = true;
+        }
+
+        tags.forEach(tag => {
+            _tags.push(tag);
+        });
+
+        return {
+            'source': path,
+            'childTags': child_tags,
+            'tags': _tags,
+            'featured': featured,
+            'thumbnail': thumbnail_path,
+            'title': title
+        };
+    },
+
+    _convert_recursive: function (dir_path, categories, output) {
+        let directory = Gio.File.new_for_path(dir_path);
+        let enumerator = directory.enumerate_children('*', Gio.FileQueryInfoFlags.NONE, null);
+        let info = enumerator.next_file(null);
+
+        while (info !== null) {
+            let file = enumerator.get_child(info);
+            let file_type = info.get_file_type();
+            let mime_type = info.get_content_type();
+            let path = file.get_path();
+            let name = info.get_name().split('.')[0];
+
+            if (file_type === Gio.FileType.DIRECTORY) {
+                let _categories = categories.slice(0);
+                _categories.push(name);
+
+                this._convert_recursive(path, _categories, output);
+
+                let tags = [];
+                if (categories.length > 0)
+                    tags = [categories[categories.length - 1]];
+
+                let thumbnail_path = path + '/thumbnail.jpg';
+
+                output['sets'].push(this._create_set(path, name, [name], tags, thumbnail_path));
+
+            } else if (mime_type.startsWith('image')) {
+                output['images'].push(this._create_image(path, mime_type));
+
+            } else if (mime_type === 'application/pdf') {
+                this._sequence += 1;
+
+                let category = null;
+                if (categories.length > 0)
+                    category = categories[categories.length - 1];
+
+                output['articles'].push(this._create_article(path, name, mime_type, category, this._sequence));
+
+            } else {
+               /* mehhh */
+            }
+
+            info = enumerator.next_file(null);
+        }
+    },
+
+    run: function () {
+        let output_dict = this._create_root();
+        let directories = [];
+
+        this._convert_recursive(this._dir_path, directories, output_dict);
+
+        return output_dict;
+    },
+});
+
+const USAGE = [
+    'usage: basin-generator <path_to_input_directory> <path_to_output_json>',
+    '',
+    'Utility for generating basin JSON files.',
+].join('\n');
+
+let main = function () {
```

Comment:
`function main() {` since otherwise it'll be an anonymous function assigned to the `main` variable